File: legacy/src/game/pinball-knight/fx/pools/blade-ring.ts
/**
 * Orbiting blades — the Blade Storm ring made visible.
 *
 * Extracted from the 1700-line `vfx.ts` when it moved into `fx/`. Behaviour is
 * unchanged — only the file boundary is new, and the shared constants it used to
 * hold privately now live in `./shared.ts` so eight modules cannot drift apart.
 */
import * as THREE from "three";
import { CAMERA_YAW, CAMERA_TILT } from "../../constants";

/**
 * ORBITING BLADES — Blade Storm, made visible.
 *
 * Every other pool here fires an EVENT that then decays. A buff that lasts five
 * seconds is a STATE, so this pool is keep-alive instead: `refresh()` places the
 * blades for this frame and re-arms a short hold; `update()` hides them once the
 * caller stops refreshing (i.e. the buff lapsed) — no teardown call to forget,
 * and a dropped frame just leaves the ring up a beat longer.
 *
 * The blades are the SLASH crescent texture on billboards, rolled in screen
 * space so each one leans along its own tangent — the same trick the melee
 * slash uses to aim a flat quad in an iso view.
 */
const BLADE_MAX = 6;
const BLADE_HOLD = 0.12; // seconds a placed blade survives without a refresh

export class BladeRing {
  readonly group: THREE.Group;
  private meshes: THREE.Mesh[] = [];
  private geo: THREE.PlaneGeometry;
  private hold = 0;

  constructor(tex: THREE.CanvasTexture) {
    this.group = new THREE.Group();
    this.geo = new THREE.PlaneGeometry(0.85, 0.85);
    for (let i = 0; i < BLADE_MAX; i++) {
      const mat = new THREE.MeshBasicMaterial({
        map: tex,
        transparent: true,
        depthWrite: false,
        depthTest: true,
        blending: THREE.AdditiveBlending,
        side: THREE.DoubleSide,
        opacity: 0.85,
      });
      const m = new THREE.Mesh(this.geo, mat);
      m.rotation.order = "YXZ";
      m.visible = false;
      m.renderOrder = 12;
      this.meshes.push(m);
      this.group.add(m);
    }
  }

  refresh(x: number, y: number, z: number, angle: number, count: number, radius: number, color: number): void {
    this.hold = BLADE_HOLD;
    const n = Math.min(count, BLADE_MAX);
    for (let i = 0; i < BLADE_MAX; i++) {
      const m = this.meshes[i];
      if (i >= n) {
        m.visible = false;
        continue;
      }
      const a = angle + (i / n) * Math.PI * 2;
      m.position.set(x + Math.cos(a) * radius, y, z + Math.sin(a) * radius);
      // Lean the crescent along the orbit tangent (screen-space roll), then
      // billboard it to the fixed iso camera like every other flat FX quad.
      m.rotation.z = -a - Math.PI / 2;
      m.rotation.y = CAMERA_YAW;
      m.rotation.x = -CAMERA_TILT;
      const mat = m.material as THREE.MeshBasicMaterial;
      mat.color.setHex(color);
      // A touch of per-blade flicker so the ring shimmers instead of reading as
      // a rigid decal spinning at a constant brightness.
      mat.opacity = 0.7 + Math.random() * 0.3;
      m.scale.setScalar(0.9 + Math.random() * 0.2);
      m.visible = true;
    }
  }

  update(dt: number): void {
    if (this.hold <= 0) return;
    this.hold -= dt;
    if (this.hold > 0) return;
    for (const m of this.meshes) m.visible = false;
  }

  /** See `warmupReveal` — slot 0 stands in for the pool at prewarm time. */
  warmupTarget(): THREE.Object3D {
    return this.meshes[0];
  }

  dispose(): void {
    this.geo.dispose();
    for (const m of this.meshes) (m.material as THREE.Material).dispose();
  }
}

File: crates/pk-core/src/fx/blade_ring.rs
//! Orbiting Blades VFX Ring Pool — Blade Storm orbiting crescent blades with keep-alive state tracking.
//!
//! PORTS: `fx/pools/blade-ring.ts`

use std::f32::consts::PI;

pub const BLADE_MAX: usize = 6;
pub const BLADE_HOLD: f32 = 0.12; // seconds a placed blade survives without a refresh
pub const CAMERA_YAW: f32 = PI / 4.0;
pub const CAMERA_TILT: f32 = 30.0 * (PI / 180.0);

#[derive(Clone, Debug, PartialEq)]
pub struct BladeInstance {
    pub pos: [f32; 3],
    pub rot_z: f32,
    pub rot_y: f32,
    pub rot_x: f32,
    pub color_hex: u32,
    pub opacity: f32,
    pub scale: f32,
    pub visible: bool,
}

impl BladeInstance {
    pub fn new() -> Self {
        Self {
            pos: [0.0, 0.0, 0.0],
            rot_z: 0.0,
            rot_y: CAMERA_YAW,
            rot_x: -CAMERA_TILT,
            color_hex: 0xffffff,
            opacity: 0.85,
            scale: 1.0,
            visible: false,
        }
    }
}

impl Default for BladeInstance {
    fn default() -> Self {
        Self::new()
    }
}

#[derive(Clone, Debug, PartialEq)]
pub struct BladeRing {
    pub meshes: Vec<BladeInstance>,
    pub hold: f32,
}

impl BladeRing {
    pub fn new() -> Self {
        let mut meshes = Vec::with_capacity(BLADE_MAX);
        for _ in 0..BLADE_MAX {
            meshes.push(BladeInstance::new());
        }
        Self { meshes, hold: 0.0 }
    }

    /// Positions orbiting blades around actor position for this frame and arms hold timer.
    pub fn refresh(
        &mut self,
        pos: [f32; 3],
        angle: f32,
        count: usize,
        radius: f32,
        color_hex: u32,
        mut rng: impl FnMut() -> f32,
    ) {
        self.hold = BLADE_HOLD;
        let n = count.min(BLADE_MAX);

        for i in 0..BLADE_MAX {
            let m = &mut self.meshes[i];
            if i >= n {
                m.visible = false;
                continue;
            }

            let a = angle + (i as f32 / n as f32) * PI * 2.0;
            m.pos = [
                pos[0] + a.cos() * radius,
                pos[1],
                pos[2] + a.sin() * radius,
            ];

            // Lean crescent along orbit tangent, then billboard to iso camera
            m.rot_z = -a - PI / 2.0;
            m.rot_y = CAMERA_YAW;
            m.rot_x = -CAMERA_TILT;

            m.color_hex = color_hex;
            m.opacity = 0.7 + rng() * 0.3;
            m.scale = 0.9 + rng() * 0.2;
            m.visible = true;
        }
    }

    /// Ticks down keep-alive hold timer and hides blades on buff expiration.
    pub fn update(&mut self, dt: f32) {
        if self.hold <= 0.0 {
            return;
        }
        self.hold -= dt;
        if self.hold <= 0.0 {
            for m in self.meshes.iter_mut() {
                m.visible = false;
            }
        }
    }

    /// Total count of currently visible blades.
    pub fn active_count(&self) -> usize {
        self.meshes.iter().filter(|m| m.visible).count()
    }
}

impl Default for BladeRing {
    fn default() -> Self {
        Self::new()
    }
}
